Floofbot/Services/RaidProtectionService.cs

```diff
@@ -0,0 +1,387 @@
+        private Dictionary<IGuild, int> numberOfJoins = new Dictionary<IGuild, int>();
+        // The max number of repeated emojis before triggering the raid protection
+        private static int maxNumberEmojis = 5;
+
+
+
+        public RaidProtectionService()
+        {
+
+        }
+        public RaidProtectionConfig GetServerConfig(IGuild guild, FloofDataContext _floofDb)
+        {
+            RaidProtectionConfig serverConfig = _floofDb.RaidProtectionConfigs.Find(guild.Id);
+            return serverConfig;
+        }
+        public async Task NotifyModerators(SocketRole modRole, ITextChannel modChannel, string message)
+        {
+            await modChannel.SendMessageAsync(modRole.Mention + " there may be a possible raid! Reason: ``" + message + "``");
+        }
+        private async void UserPunishmentTimeout(ulong userid)
+        {
+            await Task.Delay(forgivenDuration);
+            if (userPunishmentCount.ContainsKey(userid))
+                if (userPunishmentCount[userid] != 0)
+                    userPunishmentCount[userid] -= 1;
+        }
+        private async void punishedUsersTimeout(SocketUser user)
+        {
+            await Task.Delay(removePunishedUserDelay);
+            if (punishedUsers.Contains(user))
+                punishedUsers.Remove(user);
+        }
+        private async void userJoinTimeout(IGuild guild)
+        {
+            await Task.Delay(userJoinsDelay);
+            if (numberOfJoins.ContainsKey(guild))
+                if (numberOfJoins[guild] != 0)
+                    numberOfJoins[guild] -= 1;
+        }
+        private async void UserMessageCountTimeout(ulong userid)
+        {
+            await Task.Delay(durationForMaxMessages);
+            if (userMessageCount.ContainsKey(userid))
+                if (userMessageCount[userid] != 0)
+                    userMessageCount[userid] -= 1;
+        }
+        public async Task<bool> CheckUserMessageCount(SocketMessage msg)
+        {
+            if (userMessageCount.ContainsKey(msg.Author.Id))
+            {
+                userMessageCount[msg.Author.Id] += 1;
+                if (userMessageCount[msg.Author.Id] >= 5) // no more than 5 messages in time frame
+                {
+                    // add a bad boye point for the user
+                    if (userPunishmentCount.ContainsKey(msg.Author.Id))
+                    {
+                        userPunishmentCount[msg.Author.Id] += 1;
+                    }
+                    else // they were a good boye but now they are not
+                    {
+                        userPunishmentCount.Add(msg.Author.Id, 1);
+                    }
+                    // reset the count after punishing the user
+                    userMessageCount[msg.Author.Id] = 0;
+                    // we run an async task to remove their point after the specified duration
+                    await Task.Run(async () =>
```

Comment:
Optional goal: instead of having one timer for each strike point, have one timer for all strikes. That is, the timer is reset if another strike occurs before the timer finishes. This prevents users from floating at a certain strike level indefinitely (they receive strikes at the same rate they lose them).

Reply:
This is substantially harder, and I think I would prefer this left as a future PR